source/mir/random/flex/internal/area.d

```diff
@@ -0,0 +1,1024 @@
+    static if (isHat)
+        auto sh = iv.hat;
+    else
+        auto sh = iv.squeeze;
+
+    // check difference to left and right starting point
+    // sigma in the paper (1: left side, -1, right side
+    immutable S leftOrRight = (iv.rx - sh.y) > (sh.y - iv.lx) ? 1 : -1;
+    immutable S shL = sh(iv.lx);
+    immutable S shR = sh(iv.rx);
+    immutable S ivLength = iv.rx - iv.lx;
+    auto z = leftOrRight * sh.slope * ivLength;
+
+    // sh.y is the boundary point where f obtains its maximum
+
+    // specializations for T_c family (page 6, table 2 from Botts et al. (2013))
+    if (iv.c == 0)
+    {
+        if (fabs(z) < constants!S.smallExp)
+        {
+            area = exp(sh.a);
+            S t = (z * z * z) * one_div_24;
+            t += 1 + z * S(0.5) + (z * z) * one_div_6;
+            area *= t;
+            area *= ivLength;
+        }
+        else
+        {
+            area = exp(shR) - exp(shL);
+            area /= sh.slope;
+        }
+    }
+    else
+    {
+        z /= sh.a;
+        auto sgnc = copysign(S(1), iv.c);
+        if (!(sgnc * shL >= 0) || !(sgnc * shR >= 0))
+        {
+            area = S.max;
+        }
+        else if (iv.c == 1)
+        {
+            area = S(0.5) * sh.a * ivLength;
+            S t = z + 2;
+            area *= t;
+        }
+        else if (iv.c == S(-0.5))
+        {
+            if (fabs(z) < S(0.5))
+            {
+                area = 1 / (sh.a * sh.a) * ivLength / (z + 1);
+            }
+            else
+            {
+                S _l = 1 / shL;
+                S _r = 1 / shR;
+                area = (_l - _r) / sh.slope;
+            }
+        }
+        else if (iv.c == -1)
+        {
+            if (fabs(z) < constants!S.smallLog)
+            {
+                area = 1 - z * S(0.5) + z * z * one_div_3;
+                area *=  -ivLength;
+                area /= sh.a;
+            }
+            else
+            {
+                int lexp = void, rexp = void;
+                S b = frexp(-shL, lexp);
+                b /= frexp(-shR, rexp);
+                immutable S rem = log2(b);
+                area = lexp - rexp + rem;
+                area /= sh.slope;
+                area /= S(LOG2E);
+            }
+        }
+        else
+        {
+            if (fabs(sh.slope) < S(1e-10))
+            {
+                import std.math: sgn;
+                assert(sh.a * sgn(iv.c) >= 0);
+                area = flexInverse!true(sh.a, iv.c);
+                area *= ivLength;
+            }
+            else
+            {
+                // workaround for @@@BUG 16341 @@@
+                S r = antiderivative!true(shR, iv.c);
+                S l = antiderivative!true(shL, iv.c);
+                area = r - l;
+                area /= sh.slope;
+            }
+        }
+    }
+
+    // if we receive an invalid value, we require the interval to be split
+    if (!isFinite(area))
+        area = S.max;
+    else if (area < 0)
+        area = S.max;
```

Comment:
this safety check is never executed, so while being possible in theory I lack an example where it actually happens